### ISSUE TITLE
Custom Env & DQN Policy Returns the same action after 10 seconds (action can be different at each appliction run)

### ISSUE THREAD

**Issue by cosmir17** [tensorflow/agents#521]
Hi :)

I enjoy playing Nimble Quest game. 
https://store.steampowered.com/app/259780/Nimble_Quest/

I am using 'tensorflow/agents DqnAgent'.
```
q_net = q_network.QNetwork(
    nimble_quest_env.observation_spec(),
    nimble_quest_env.action_spec(),
    # dropout_layer_params=dropout_layer,
    conv_layer_params=conv_layer_params,
    fc_layer_params=fc_layer_params
)

optimizer = tf.compat.v1.train.AdamOptimizer(learning_rate=learning_rate)
train_step_counter = tf.Variable(0)

agent = dqn_agent.DqnAgent(
    nimble_quest_env.time_step_spec(),
    nimble_quest_env.action_spec(),
    epsilon_greedy=0.1,
    q_network=q_net,
    optimizer=optimizer,
    td_errors_loss_fn=common.element_wise_squared_loss,
    train_step_counter=train_step_counter)

agent.initialize()
```

I am creating A.I bot to play the game. My TF DQN agent yields just one action after 10 seconds it's running! Kindly note that it can be a different action at each application run.
![Screenshot 2020-11-23 at 18 45 56](https://user-images.githubusercontent.com/2845708/100003137-821dec00-2dbd-11eb-96c3-e43c99771e43.png)
![Screenshot 2020-11-23 at 18 46 04](https://user-images.githubusercontent.com/2845708/100003204-9b269d00-2dbd-11eb-8552-b105156a29ef.png)

I have tried with 0.6.0 and tf-agent-nightly versions. Both shows the same result. 
I spent three days but can't figure out what's wrong with my code.
Random policy is actually much better.

I found similar issues but this issue is not related.
https://github.com/tensorflow/agents/issues/258
https://github.com/tensorflow/agents/issues/161

main file:
https://github.com/cosmir17/nimble-quest-tf-agent/blob/master/main_without_save_point.py

custom environment file:
https://github.com/cosmir17/nimble-quest-tf-agent/blob/master/nq_environment.py

Can you have a look and find out what I am missing, please?

**Comment by sguada** [comment]
You may need to collect more initial data with the random policy and annile the epsilon as training advances.

You probably want to make sure to look at the TensorBoard and enable debug summaries to make sure learning is happening, and losses are evolving.

```
epsilon = tf.compat.v1.train.polynomial_decay(
            1.0,
            train_step_counter,
            epsilon_decay_period,
            end_learning_rate=0.1)
```

**Comment by cosmir17** [comment]
Hi @sguada,
I set 'initial_collect_steps' to be 10000 and ran for one hour to get RandomPolicyAgent data but the result was same.
Also, I changed 'epsilon_greedy' value in DqnAgent(...) to 0.3, 0.4, 0.8 but it didn't affect the result.

I spent three days (8h each day) and I have no idea what's wrong with my code.

Somehow, it doesn't work for image observation spec.. Do I need to do extra for image fed RL? my conv_layer_params for QNetwork(...) might be wrong? I also looked at the TF library test classes but the only thing I saw different was 'output_dim' setting.. but I don't think that's needed for my case.

**Comment by sguada** [comment]
I'm sorry but debugging somebody else code is very difficult. 

My advice is to enable debug_summaries to monitor how the losses, weights, gradients evolve overtime, and upload those to https://tensorboard.dev/ using `tensorboard dev upload --logdir=./logs/`

For instance just skimming your code it seems that you use `batch_size=1` (You will need at least `batch_size=32`) to learn better.
Another thing I notice is that the images you are using are `(250, 250, 3)` which can be quite big, maybe try resizing to a smaller image, and consider using some the [atari_preprocessing](https://github.com/tensorflow/agents/blob/master/tf_agents/environments/atari_preprocessing.py)

**Comment by cosmir17** [comment]
Although it's not written by a Python dev(myself), I wrote them based on 
https://github.com/tensorflow/agents/blob/master/docs/tutorials/1_dqn_tutorial.ipynb
https://github.com/tensorflow/agents/blob/master/docs/tutorials/2_environments_tutorial.ipynb

The only difference here is I feed image.
Thank you for the Atari example!! I shall adapt some techniques there.

I saw your advice and used '1' as batch size for I was speculating tf-wrapper was the cause of this problem. It turned out it's not for my case as you explained here.
https://github.com/tensorflow/agents/issues/258#issuecomment-567811189

After increasing the batch size. I saw an improvement(small but it's something). I shall experiment.

**Comment by cosmir17** [comment]
Hello @sguada or anyone interested,
I made some improvements (rgb => gray, better score recognition etc) 
https://github.com/cosmir17/nimble-quest-tf-agent/blob/master/main.py

I saw TF agents Atari and CartPole examples. The gameplays are accelerated. So, the RL training can be done within 10 to 20 minutes, and the results are promising.

Nonetheless, I have been training an agent with a real game. My agent read the screen, and the code presses appropriate keys. The training process takes a too long time (e.g. 4 hours for 30 iterations) as I can't accelerate the actual gameplay. It doesn't yield a satisfactory result. It's rather a disappointment..

![Screenshot 2020-11-27 at 18 09 57](https://user-images.githubusercontent.com/2845708/100475793-ebfb0600-30db-11eb-8cdc-e979bdb0b2e0.png)

I need at least 500 iterations to see some decent result. I don't think even 24 hours will be sufficient to reach a normal level.

May I ask if the following options are possible?
1. Is it possible to run the code and the game inside VM with an accelerated timeclock? Consequently, the agent will play fast.
2. Is there a python library that does containerisation for RL training? 

Thank you,
Sean
